File: core/src/nagma_core/compiler.py
"""Compiler: NagmaDoc + (BPM, Sa) -> ExpressiveScore.

This is where the two non-negotiable invariants are enforced:

  1. Machine-perfect laya. matra_dur_s = 60/bpm exactly; every matra boundary
     and sam sits on an exact multiple of it. loop_length_samples is an exact
     integer so the on-device loop restarts with zero drift.
  2. Human timbre via pure code. The performance model perturbs only
     non-structural onsets and shapes dynamics/legato/per-avartan variation.

The compiler expands the one-avartan NagmaDoc into a super-loop of `avartans`
cycles, each with its own RNG seed.
"""

from __future__ import annotations

from .models import Event, ExpressiveScore, NagmaDoc
from . import performance as perf
from .reduce import laya_for_bpm, reduce_doc
from .sargam import note_to_midi, sa_to_midi
from .taal import get_taal

DEFAULT_SAMPLE_RATE = 44100
DEFAULT_AVARTANS = 4


def _one_avartan_events(
    doc: NagmaDoc, matra_dur_s: float, sa_midi: int, marks: list[str]
) -> list[dict]:
    """Build the base (jitter-free, base-velocity) events for a single cycle.

    Sustains fold into the preceding note's duration. Returns plain dicts with
    the fields the expansion pass needs.
    """
    events: list[dict] = []
    last: dict | None = None  # the most recent struck note, for sustain folding

    for matra in doc.matras:
        n = len(matra.notes)
        # Slot width: whole matra for a single note, half each for a pair.
        for slot, note in enumerate(matra.notes):
            slot_dur = matra_dur_s / n
            slot_start = matra.index * matra_dur_s + slot * slot_dur
            # The first slot of a matra lands on a matra boundary -> structural.
            structural = slot == 0

            if note.kind == "sustain":
                if last is not None:
                    last["dur_s"] += slot_dur
                # A sustain with no predecessor is silence (parser forbids it
                # at sam; mid-piece it just yields a rest).
                continue

            midi = note_to_midi(note.swar, note.octave, sa_midi)
            ev = {
                "start_s": slot_start,
                "dur_s": slot_dur,
                "midi": midi,
                "matra": matra.index,
                "structural": structural,
                "mark": marks[matra.index],
            }
            events.append(ev)
            last = ev

    return events


def compile_score(
    doc: NagmaDoc,
    *,
    bpm: float,
    sa: str | int,
    sample_rate: int = DEFAULT_SAMPLE_RATE,
    avartans: int = DEFAULT_AVARTANS,
    seed: int = 0,
    instrument: str = "harmonium",
    grace_density: float | None = None,
) -> ExpressiveScore:
    """Compile a NagmaDoc into an ExpressiveScore super-loop."""
    if grace_density is None:
        grace_density = perf.GRACE_DENSITY
    if bpm <= 0:
        raise ValueError("bpm must be positive")
    if avartans < 1:
        raise ValueError("avartans must be >= 1")

    taal_def = get_taal(doc.taal)
    marks = taal_def.matra_marks()
    matra_count = taal_def.matra_count
    if len(doc.matras) != matra_count:
        raise ValueError(
            f"NagmaDoc has {len(doc.matras)} matras but {taal_def.name} "
            f"requires {matra_count}."
        )

    sa_midi = sa if isinstance(sa, int) else sa_to_midi(sa)

    # --- exact timing (invariant #1) --------------------------------------
    matra_dur_s = 60.0 / bpm
    avartan_dur_s = matra_count * matra_dur_s
    loop_length_s = avartans * avartan_dur_s
    loop_length_samples = round(loop_length_s * sample_rate)

    base = _one_avartan_events(doc, matra_dur_s, sa_midi, marks)

    # Add legato overlap once on the base durations.
    for ev in base:
        ev["dur_s"] += perf.LEGATO_OVERLAP_S

    # --- expand across avartans with per-cycle variation (invariant #2) ---
    pitch_set = sorted({ev["midi"] for ev in base})

    events: list[Event] = []
    for a in range(avartans):
        rng = perf.avartan_rng(seed, a)
        grng = perf.grace_rng(seed, a)  # separate stream: graces don't perturb feel
        cycle_offset = a * avartan_dur_s
        for ev in base:
            jitter = perf.timing_jitter(rng, ev["structural"])
            start = cycle_offset + ev["start_s"] + jitter
            vel = perf.base_velocity(ev["matra"], ev["mark"], matra_count)
            vel = perf.apply_velocity_noise(rng, vel)
            swell = perf.swell_depth(rng, ev["dur_s"], ev["structural"])
            events.append(
                Event(
                    start_s=start,
                    dur_s=ev["dur_s"],
                    midi=ev["midi"],
                    velocity=vel,
                    matra=ev["matra"],
                    avartan=a,
                    structural=ev["structural"],
                    swell=swell,
                )
            )
            # Kan swar: sparingly flick a neighbour swar in just before this note.
            # It steals time from BEFORE the exact onset, so the grid is untouched.
            if ev["dur_s"] >= perf.GRACE_MIN_MAIN_S and grng.random() < grace_density:
                kan = perf.kan_pitch(ev["midi"], pitch_set)
                gstart = start - perf.GRACE_DUR_S
                if kan is not None and kan != ev["midi"] and gstart >= cycle_offset:
                    events.append(
                        Event(
                            start_s=gstart,
                            dur_s=perf.GRACE_DUR_S + perf.GRACE_LEGATO_S,
                            midi=kan,
                            velocity=max(1, int(vel * perf.GRACE_VEL_SCALE)),
                            matra=ev["matra"],
                            avartan=a,
                            structural=False,
                            swell=0.0,
                        )
                    )

    events.sort(key=lambda e: e.start_s)

    # Same-pitch re-articulation: legato must not sustain a note into the next
    # onset of the SAME pitch (that retriggers the sampler mid-note and the
    # repeat sounds cut off — e.g. "Sa Sa Sa" blurs into one). Clamp each note
    # to end a small gap before the next same-pitch onset. Different-pitch
    # overlap (the harmonium bleed) is left untouched.
    by_pitch: dict[int, list[Event]] = {}
    for e in events:
        by_pitch.setdefault(e.midi, []).append(e)
    for group in by_pitch.values():
        for a, b in zip(group, group[1:]):  # already start-sorted (events were)
            latest_end = b.start_s - perf.REARTICULATION_GAP_S
            if a.start_s + a.dur_s > latest_end:
                a.dur_s = max(perf.MIN_NOTE_S, latest_end - a.start_s)

    return ExpressiveScore(
        sample_rate=sample_rate,
        bpm=bpm,
        sa_midi=sa_midi,
        instrument=instrument,
        avartans=avartans,
        matra_count=matra_count,
        matra_dur_s=matra_dur_s,
        avartan_dur_s=avartan_dur_s,
        loop_length_s=loop_length_s,
        loop_length_samples=loop_length_samples,
        events=events,
        seed=seed,
        bellows=perf.bellows_params(),
    )


def realize(
    doc: NagmaDoc,
    *,
    bpm: float,
    sa: str | int,
    laya: str | None = None,
    sample_rate: int = DEFAULT_SAMPLE_RATE,
    avartans: int = DEFAULT_AVARTANS,
    seed: int = 0,
    instrument: str = "harmonium",
) -> ExpressiveScore:
    """Author-once entry point: pick a laya from `bpm`, REDUCE the authored
    (vilambit) doc to that laya, then compile — with ornament density scaled to
    the laya. Pass `laya` to override the automatic BPM->laya selection.
    """
    if laya is None:
        laya = laya_for_bpm(bpm)
    reduced = reduce_doc(doc, laya)
    return compile_score(
        reduced,
        bpm=bpm,
        sa=sa,
        sample_rate=sample_rate,
        avartans=avartans,
        seed=seed,
        instrument=instrument,
        # Plucked instruments (sitar) take no kan-swar graces.
        grace_density=0.0
        if perf.is_plucked(instrument)
        else perf.GRACE_DENSITY_BY_LAYA.get(laya, perf.GRACE_DENSITY),
    )

File: app/lib/render/compiler.dart
// Compiler: NagmaDoc + (BPM, Sa) -> ExpressiveScore.
//
// Dart port of core/src/nagma_core/compiler.py. Enforces the two invariants:
//   1. Machine-perfect laya. matraDurS = 60/bpm exactly; every matra boundary and
//      sam sits on an exact multiple of it. loopLengthSamples is an exact integer.
//   2. Human timbre via pure code. The performance model perturbs only
//      non-structural onsets and shapes dynamics/legato/per-avartan variation.

import 'dart:math' as math;

import 'models.dart';
import 'performance.dart' as perf;
import 'reduce.dart';
import 'sargam.dart';
import 'taal.dart';

const int defaultSampleRate = 44100;
const int defaultAvartans = 4;

/// Mutable base (jitter-free, base-velocity) event for a single cycle.
class _BaseEvent {
  double startS;
  double durS;
  final int midi;
  final int matra;
  final bool structural;
  final String mark;
  _BaseEvent({
    required this.startS,
    required this.durS,
    required this.midi,
    required this.matra,
    required this.structural,
    required this.mark,
  });
}

List<_BaseEvent> _oneAvartanEvents(
  NagmaDoc doc,
  double matraDurS,
  int saMidi,
  List<String> marks,
) {
  final events = <_BaseEvent>[];
  _BaseEvent? last; // the most recent struck note, for sustain folding

  for (final matra in doc.matras) {
    final n = matra.notes.length;
    for (var slot = 0; slot < n; slot++) {
      final note = matra.notes[slot];
      final slotDur = matraDurS / n;
      final slotStart = matra.index * matraDurS + slot * slotDur;
      // The first slot of a matra lands on a matra boundary -> structural.
      final structural = slot == 0;

      if (note.kind == 'sustain') {
        if (last != null) last.durS += slotDur;
        // A sustain with no predecessor is silence (parser forbids it at sam).
        continue;
      }

      final midi = noteToMidi(note.swar!, note.octave, saMidi);
      final ev = _BaseEvent(
        startS: slotStart,
        durS: slotDur,
        midi: midi,
        matra: matra.index,
        structural: structural,
        mark: marks[matra.index],
      );
      events.add(ev);
      last = ev;
    }
  }

  return events;
}

/// Compile a NagmaDoc into an ExpressiveScore super-loop.
ExpressiveScore compileScore(
  NagmaDoc doc, {
  required double bpm,
  required Object sa, // String key or int MIDI
  int sampleRate = defaultSampleRate,
  int avartans = defaultAvartans,
  int seed = 0,
  String instrument = 'harmonium',
  double? graceDensity,
}) {
  graceDensity ??= perf.graceDensity;
  if (bpm <= 0) throw ArgumentError('bpm must be positive');
  if (avartans < 1) throw ArgumentError('avartans must be >= 1');

  final taalDef = getTaal(doc.taal);
  final marks = taalDef.matraMarks();
  final matraCount = taalDef.matraCount;
  if (doc.matras.length != matraCount) {
    throw ArgumentError(
      'NagmaDoc has ${doc.matras.length} matras but ${taalDef.name} '
      'requires $matraCount.',
    );
  }

  final saMidi = sa is int ? sa : saToMidi(sa as String);

  // --- exact timing (invariant #1) --------------------------------------
  final matraDurS = 60.0 / bpm;
  final avartanDurS = matraCount * matraDurS;
  final loopLengthS = avartans * avartanDurS;
  final loopLengthSamples = (loopLengthS * sampleRate).round();

  final base = _oneAvartanEvents(doc, matraDurS, saMidi, marks);

  // Add legato overlap once on the base durations.
  for (final ev in base) {
    ev.durS += perf.legatoOverlapS;
  }

  // --- expand across avartans with per-cycle variation (invariant #2) ---
  final pitchSet = (base.map((e) => e.midi).toSet().toList())..sort();

  final plucked = perf.isPlucked(instrument);
  final events = <Event>[];
  for (var a = 0; a < avartans; a++) {
    final rng = perf.avartanRng(seed, a);
    final grng = perf.graceRng(seed, a); // separate stream
    final prng = perf.pluckRng(seed, a); // plucked expression stream
    final cycleOffset = a * avartanDurS;
    int? prevMidi; // previous struck pitch this avartan, for meend
    for (final ev in base) {
      final jitter = plucked
          ? perf.timingJitterGaussian(rng, ev.structural)
          : perf.timingJitter(rng, ev.structural);
      final start = cycleOffset + ev.startS + jitter;
      var vel = perf.baseVelocity(ev.matra, ev.mark, matraCount);
      vel = plucked
          ? perf.applyVelocityNoisePlucked(rng, vel)
          : perf.applyVelocityNoise(rng, vel);
      // A pluck attacks then decays — no reed-style swell. Instead give it
      // per-note micro-detune (round-robin) and meend glides into the note.
      final swell = plucked ? 0.0 : perf.swellDepth(rng, ev.durS, ev.structural);
      double microCents = 0.0;
      int? glideFrom;
      double glideS = 0.0;
      if (plucked) {
        microCents = perf.microDetuneCents(prng);
        if (ev.durS >= perf.meendMinNoteS) {
          glideFrom = perf.meendFrom(prng, prevMidi, ev.midi, ev.structural);
          if (glideFrom != null) glideS = math.min(perf.meendGlideS, 0.6 * ev.durS);
        }
      }
      prevMidi = ev.midi;
      events.add(Event(
        startS: start,
        durS: ev.durS,
        midi: ev.midi,
        velocity: vel,
        matra: ev.matra,
        avartan: a,
        structural: ev.structural,
        swell: swell,
        microCents: microCents,
        glideFromMidi: glideFrom,
        glideS: glideS,
      ));
      // Kan swar: sparingly flick a neighbour swar in just before this note.
      // It steals time from BEFORE the exact onset, so the grid is untouched.
      if (ev.durS >= perf.graceMinMainS && grng.random() < graceDensity) {
        final kan = perf.kanPitch(ev.midi, pitchSet);
        final gstart = start - perf.graceDurS;
        if (kan != null && kan != ev.midi && gstart >= cycleOffset) {
          events.add(Event(
            startS: gstart,
            durS: perf.graceDurS + perf.graceLegatoS,
            midi: kan,
            velocity: math.max(1, (vel * perf.graceVelScale).toInt()),
            matra: ev.matra,
            avartan: a,
            structural: false,
            swell: 0.0,
          ));
        }
      }
    }
  }

  events.sort((a, b) => a.startS.compareTo(b.startS));

  // Same-pitch re-articulation: legato must not sustain a note into the next
  // onset of the SAME pitch. Clamp each note to end a small gap before the next
  // same-pitch onset. Different-pitch overlap (harmonium bleed) is left alone.
  final byPitch = <int, List<Event>>{};
  for (final e in events) {
    byPitch.putIfAbsent(e.midi, () => []).add(e);
  }
  for (final group in byPitch.values) {
    for (var i = 0; i + 1 < group.length; i++) {
      final a = group[i];
      final b = group[i + 1];
      final latestEnd = b.startS - perf.rearticulationGapS;
      if (a.startS + a.durS > latestEnd) {
        a.durS = math.max(perf.minNoteS, latestEnd - a.startS);
      }
    }
  }

  return ExpressiveScore(
    sampleRate: sampleRate,
    bpm: bpm,
    saMidi: saMidi,
    instrument: instrument,
    avartans: avartans,
    matraCount: matraCount,
    matraDurS: matraDurS,
    avartanDurS: avartanDurS,
    loopLengthS: loopLengthS,
    loopLengthSamples: loopLengthSamples,
    events: events,
    seed: seed,
    bellows: perf.bellowsParams(),
  );
}

/// Author-once entry point: pick a laya from `bpm`, REDUCE the authored
/// (vilambit) doc to that laya, then compile — with ornament density scaled to
/// the laya. Pass `laya` to override the automatic BPM->laya selection.
ExpressiveScore realize(
  NagmaDoc doc, {
  required double bpm,
  required Object sa,
  String? laya,
  int sampleRate = defaultSampleRate,
  int avartans = defaultAvartans,
  int seed = 0,
  String instrument = 'harmonium',
}) {
  laya ??= layaForBpm(bpm);
  final reduced = reduceDoc(doc, laya);
  return compileScore(
    reduced,
    bpm: bpm,
    sa: sa,
    sampleRate: sampleRate,
    avartans: avartans,
    seed: seed,
    instrument: instrument,
    // Plucked instruments (sitar) take no kan-swar graces.
    graceDensity: perf.isPlucked(instrument)
        ? 0.0
        : (perf.graceDensityByLaya[laya] ?? perf.graceDensity),
  );
}
